src/rust/engine/src/rule_graph.rs

```diff
@@ -508,9 +556,53 @@ pub struct RuleGraph {
 }
 
 impl RuleGraph {
+  pub fn print_errors(&self, externs: &Externs) {
+    // TODO the rule display is really unfriendly right now. Next up should be to improve it.
+    let mut collated_errors: HashMap<Task, HashMap<String, HashSet<TypeId>>> = HashMap::new();
+
+    let used_rules: HashSet<_> = self.rule_dependency_edges.keys().map(|entry| &entry.rule).collect();
+    for (rule_entry, diagnostics) in self.unfulfillable_rules.iter() {
+      match rule_entry {
+        &Entry::InnerEntry(ref inner) => {
+          if used_rules.contains(&inner.rule) {
+            continue
+          }
+          for d in diagnostics.iter() {
+            let mut msg_to_type = collated_errors.entry(inner.rule.clone())
+              .or_insert(HashMap::new());
+            let mut subject_set = msg_to_type.entry(d.reason.clone())
+              .or_insert(HashSet::new());
+            subject_set.insert(d.subject_type.clone());
+          }
+        },
+        _ => { } // We're only checking rule usage not entry usage generally.
+                 // So we ignore entries that do not have rules.
+      }
+    }
+    fn format_msgs(externs: &Externs, rule: &Task, subject_types_by_reasons: &HashMap<String, HashSet<TypeId>>) -> String {
+      let mut errors: Vec<_> = subject_types_by_reasons.iter().map(|(reason, subject_types)|
+        format!("{} with subject types: {}",
+                reason,
+                subject_types.iter().map(|t| externs.id_to_str(t.0)).collect::<Vec<String>>().join(", "))
+      ).collect();
+      errors.sort();
+      format!("{:?}: {}", rule, errors.join("\n    "))
+    }
+    let mut msgs: Vec<String> = collated_errors.into_iter()
+      .map(|(ref rule, ref subject_types_by_reasons)| format_msgs(externs, rule, subject_types_by_reasons))
+      .collect();
+    msgs.sort();
+
+    println!("Rules with errors {}:\n  {}", msgs.len(), msgs.join("  \n"));
+  }
+
   pub fn has_errors(&self) -> bool {
```

Comment:
Rather than `has_errors() -> bool`, doing something like `validate() -> Result<(),String>` would make more sense to me: that is, "the result is either unit because it succeeded, or a String because it failed.